### PR TITLE
Issue 159: Add common.label and common.annotations Helm parameters

### PR DESCRIPTION
**Pull request by Viktorsubota**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have bumped the version number in `/charts/flagsmith/Chart.yaml` in the section `version` or I'm merging to a
      release branch

### Features
* Adding `commonLabels` and `commonAnnotations` field to the values file
* Updating `flagsmith.labels` to use `commonLabels` if provided
* Adding `flagsmith.annotations` to use `commonannotations` if provided
* Adding annotations to places where they were missed
* Update existing annotations to add `commonAnnotations` as well

### Fixes
* Fix wrong annotations used in `service-fronend.yaml` file

## How did you test this code?

I ran `helm install` with different parameters to verify that new labels and annotations are added only if needed.

### CODE REVIEW

File: charts/flagsmith/templates/_helpers.tpl [hunk 1]
```diff
@@ -51,6 +54,25 @@ app.kubernetes.io/name: {{ include "flagsmith.name" . }}
...
+{{- end }}
+{{- with .context.customAnnotations }}
+{{ . | toYaml }}
+{{- end }}
+{{- else -}}
+{{- with .Values.commonAnnotations }}
+{{ . | toYaml }}
+{{- end }}
+{{- end }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (Viktorsubota) [reply charts/flagsmith/templates/_helpers.tpl:74]: This piece of code is designed to include annotations from the `commonAnnotations` section in the values file. If the "context" parameter is not set to the default value (.), but instead is a map containing two keys (`customAnnotations` and `context`), the code includes both the provided annotations and those from `commonAnnotations`. This flexibility is useful when different sets of values need to be passed.

The "context" key in this code is not a standard Helm feature for passing values. It's used here as a parameter to conditionally determine whether to include additional annotations. For instance, when you have different values for specific resources, such as `.Values.ingress.api.annotations` for ingress resources or `.Values.service.api.annotations` for services.

Helm/Go templates have some limitations in terms of flexibility, and this code is a workaround to handle different annotation scenarios.

A similar approach can be found in Bitnami Helm Charts, [here](https://github.com/bitnami/charts/blob/9fab9f9c645e05d79c31172a36539d2a4fcf9a44/bitnami/common/templates/_labels.tpl#L2).

[the remark this answers is not in the document]
Comment (Viktorsubota) [reply charts/flagsmith/templates/_helpers.tpl:74]: Hey @matthewelwell, any updates on this?

[the remark this answers is not in the document]
Comment (matthewelwell) [reply charts/flagsmith/templates/_helpers.tpl:74]: Hi @Viktorsubota, apologies for the delay here, I've been a bit swamped with other responsibilities. 

Given your explanation here, I would expect to see some reference to the `context` in the docs changes you added in the PR here https://github.com/Flagsmith/flagsmith/pull/3303 but perhaps I'm still misunderstanding? Are you able to provide some kind of example snippet from a `values.yaml` where the context attribute is used?

[the remark this answers is not in the document]
Comment (Viktorsubota) [reply charts/flagsmith/templates/_helpers.tpl:74]: Hello @matthewelwell , 

Let me explain a bit differently. 

I added `commonAnnotations` and `commonLabels` to be attached to all resources crated by the chart. But there is a collision with the resources "annotated" with their specific annotations.

So the idea here is to pass existing annotations to the template so It can render both if anythig is passed, or just `commonAnnotaions`. 
Here is an example for the frontend service resource:
```
 {{- $annotations := include "flagsmith.annotations" ( dict "customAnnotations" .Values.service.frontend.annotations "context" . ) }}
```

What's exectly happeneding here is `flagsmith.annotations` is called with a dict parameter (Helm donsn't support passing multiple parameters here) with 2 keys: 
* `customAnnotations: Resource-specific annotations.
 * `context: The parameter which we usually pass to the include function. It's needed to access .Values.commonAnnotations within the template.

The logic in `flagsmith.annotations` checks if `.` is a dict with those parameters to render both. Or renders just `commonAnnotations` from `.` (we pass `.` by default). 

**Note**, regardless of what you pass to include, the parameter name is always . within a template, and . as an argument is not equal to . as a parameter inside the template..

To summarize:
There were a few places where annotations were already used. In order to avoid duplicating templates, I added some logic to receive and render these annotations alongside commonAnnotations.

Please let me know if this addresses all your questions.
Also, I'm planning to update the documentation PR to include this.

Thanks!

[the remark this answers is not in the document]
Comment (Viktorsubota) [reply charts/flagsmith/templates/_helpers.tpl:74]: Oh, I just looked through the Helm doc in the main repo one more. I don't think including information about `context` and other internals of the chart would be beneficial there.

Adding annotations to resources is a one-time afford done by devs, but the docs are focused on use by users.

If you believe including information about the implementation would help, let me know, and I will update the doc.

[the remark this answers is not in the document]
Comment (khvn26) [reply charts/flagsmith/templates/_helpers.tpl:74]: Additionally, we need to take care of:
- The currently failing linting
- Rebasing/merging in the main branch

**Comment by Viktorsubota** [comment]
@matthewelwell , @khvn26 , hey guys, sorry for a late fix - a lot of work.
All fixed, please take a look.

**Review by khvn26 — APPROVED** [review]
LGTM 👍
